twpca/crossval.py

```diff
@@ -0,0 +1,212 @@
+        test_idx = idx[r]
+        yield train_idx, test_idx    
+
+def cross_validate(model, data, method, K, max_fits=np.inf, seed=1234, **fit_kw):
+    """Runs specified cross-validation method.
+
+    Args:
+        model: TWPCA model instance
+        data: array-like (n_trials x n_timepoints x n_neurons)
+        method: string specifying method {'kfold', 'leaveout'}
+        K: int, crossvalidation parameter
+
+    Keyword Args:
+        max_fits: int, maximum number of partitions to train on
+        seed: int, passed to numpy.random.seed()
+        **fit_kw: additional keywords passed to model.fit
+
+    Returns:
+        results: dict, model parameters and metrics calculated for all sessions
+    """
+
+    # set random state
+    np.random.seed(seed)
+
+    # set up cross validation partitions
+    if method == 'kfold':
+        partitions = k_fold_iter(data.shape[2], K)
+    elif method == 'leaveout':
+        partitions = leave_k_out_iter(data.shape[2], K)
+    else:
+        raise ValueError('Cross-validation method not recognized.')
+
+    nfits = 0
+    results = []
+    for train, test in partitions:
+        # partition dataset
+        traindata = data[:, :, train]
+        testdata = data[:, :, test]
+
+        # fit model to training set
+        tf.reset_default_graph() # TODO (ben): better session management
+        sess = tf.Session()
+        model.fit(traindata, sess=sess, **fit_kw)
+
+        # assess dimensionality of warped vs unwarped testdata
+        warped_testdata = model.transform(testdata)
+        warped_rank, unwarped_rank = [], []
+        for n in range(testdata.shape[2]):
+            warped_rank.append(stable_rank(warped_testdata[:, :, n]))
+            unwarped_rank.append(stable_rank(testdata[:, :, n]))
+
+        # evaluate test error
+        X_pred = model.predict(testdata)
+        test_error = np.mean((testdata - X_pred)**2)
+
+        # compile results
+        results.append(
+            {'params': model.params,
+             'warped_rank': np.array(warped_rank),
+             'unwarped_rank': np.array(unwarped_rank),
+             'test_idx': test,
+             'train_idx': train,
+             'train_error': model._sess.run(model.recon_cost),
+             'test_error': test_error,
+             'obj_history': model.obj_history
+            }
+        )
+
+        # clean-up tensorflow
+        sess.close()
+
+        # terminate early if user is impatient
+        nfits += 1
+        if nfits >= max_fits:
+            break
+
+    return results
+
+def hyperparam_search(data, n_components, warp_scales, time_scales,
+                      warp_reg=None, time_reg=None,
+                      crossval_method='kfold', K=5, max_fits=np.inf,
+                      fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                      **model_kw):
+    """Performs cross-validation over number of components, warp regularization scale, and
+    temporal regularization scale.
+
```

Comment:
can you document somewhere how the keywoard arguments are combined? typically these kind of functions take the outer product over all parameters, but here you're taking them both in as lists of parameter settings to try